3rd-party/prrte/docs/man/man1/prte_info.1.rst
.. _man1-prte_info:

prte_info
=========

prte_info |mdash| Provide detailed information on your PRRTE installation

SYNOPSIS
--------

.. code:: sh

   prte_info ...options...

DESCRIPTION
-----------

``prte_info`` provide detailed information on your PMIx Reference Run
Time (PRRTE) installation.

Extensive help documentation for this command is provided through
``prte_info --help [topic]``.

At least for now, that content is not available in man page form.
Pull requests to add all the content (via repeatable infrastructure)
would be greatly appreciated.

.. seealso::
   :ref:`prte(1) <man1-prte>`


NEWS
Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
                        University Research and Technology
                        Corporation.  All rights reserved.
Copyright (c) 2004-2006 The University of Tennessee and The University
                        of Tennessee Research Foundation.  All rights
                        reserved.
Copyright (c) 2004-2008 High Performance Computing Center Stuttgart,
                        University of Stuttgart.  All rights reserved.
Copyright (c) 2004-2006 The Regents of the University of California.
                        All rights reserved.
Copyright (c) 2006-2023 Cisco Systems, Inc.  All rights reserved.
Copyright (c) 2006      Voltaire, Inc. All rights reserved.
Copyright (c) 2006      Sun Microsystems, Inc.  All rights reserved.
                        Use is subject to license terms.
Copyright (c) 2006-2018 Los Alamos National Security, LLC.  All rights
                        reserved.
Copyright (c) 2010-2020 IBM Corporation.  All rights reserved.
Copyright (c) 2012      Oak Ridge National Labs.  All rights reserved.
Copyright (c) 2012      Sandia National Laboratories.  All rights reserved.
Copyright (c) 2012      University of Houston. All rights reserved.
Copyright (c) 2013      NVIDIA Corporation.  All rights reserved.
Copyright (c) 2013-2018 Intel, Inc. All rights reserved.
Copyright (c) 2018-2023 Amazon.com, Inc. or its affiliates.  All Rights
                        reserved.
Copyright (c) 2019-2021 Triad National Security, LLC. All rights
                        reserved.
Copyright (c) 2021      Google, LLC. All rights reserved.
Copyright (c) 2023      Jeffrey M. Squyres.  All rights reserved.
$COPYRIGHT$

Additional copyrights may follow

$HEADER$

===========================================================================

This file contains the main features as well as overviews of specific
bug fixes (and other actions) for each version of Open MPI since
version 1.0.

As more fully described in the "Software Version Number" section in
the README file, Open MPI typically releases two separate version
series simultaneously.  Since these series have different goals and
are semi-independent of each other, a single NEWS-worthy item may be
introduced into different series at different times.  For example,
feature F was introduced in the vA.B series at version vA.B.C, and was
later introduced into the vX.Y series at vX.Y.Z.

The first time feature F is released, the item will be listed in the
vA.B.C section, denoted as:

   (** also to appear: X.Y.Z) -- indicating that this item is also
                                 likely to be included in future release
                                 version vX.Y.Z.

When vX.Y.Z is later released, the same NEWS-worthy item will also be
included in the vX.Y.Z section and be denoted as:

   (** also appeared: A.B.C)  -- indicating that this item was previously
                                 included in release version vA.B.C.

4.1.6 -- September, 2023
---------------------

- Fix configure issue with XCode 15.
- Update embedded PMIx to 3.2.5.  PMIx 3.2.5 addresses CVE-2023-41915.
  Note that prior versions of Open MPI (and their associated PMIx
  implementations) are not impacted by this CVE, because Open MPI
  never uses escalated privileges on behalf of an unprivileged user.
  We are backporting this change both because it is low risk and to
  avoid alarms from CVE scanners.
- Fix issue with buffered sends and MTL-based interfaces (Libfabric,
  PSM, Portals).
- Add missing MPI_F_STATUS_SIZE to mpi.h.  Thanks to @jprotze for
  reporting the issue.
- Update Fortran mpi module configure check to be more correct.
  Thanks to Sergey Kosukhin for identifying the issue and supplying
  the fix.
- Update to properly handle PMIx v>=4.2.3.  Thanks to Bruno Chareyre,
  Github user @sukanka, and Christof Koehler for raising the
  compatibility issues and helping test the fixes.
- Fix minor issues and add some minor performance optimizations with
  OFI support.
- Support the "striping_factor" and "striping_unit" MPI_Info names
  recomended by the MPI standard for parallel IO.
- Fixed some minor issues with UCX support.
- Minor optimization for 0-byte MPI_Alltoallw (i.e., make it a no-op).


4.1.5 -- February, 2023
-----------------------

- Fix crash in one-sided applications for certain process layouts.
- Update embedded OpenPMIx to version 3.2.4
- Fix issue building with ifort on MacOS.
- Backport patches to Libevent for CVE-2016-10195, CVE-2016-10196, and
  CVE-2016-10197.  Note that Open MPI's internal libevent does not
  use the impacted portions of the Libevent code base.
- SHMEM improvements:
  - Fix initializer bugs in SHMEM interface.
  - Fix unsigned type comparisons generating warnings.
  - Fix use after clear issue in shmem_ds_reset.
- UCX improvements
  - Fix memory registration bug that could occur when UCX was built
    but not selected.
  - Reduce overhead of add_procs with intercommunicators.
  - Enable multi_send_nb by default.
  - Call opal_progress while waiting for a UCX fence to complete.
- Fix data corruption bug in osc/rdma component.
- Fix overflow bug in alltoall collective
- Fix crash when displaying topology.
- Add some MPI_F_XXX constants that were missing from mpi.h.
- coll/ucc bug fixes.


4.1.4 -- May, 2022
------------------

- Fix possible length integer overflow in numerous non-blocking collective
  operations.
- Fix segmentation fault in UCX if MPI Tool interface is finalized before
  MPI_Init is called.
- Remove /usr/bin/python dependency in configure.
- Fix OMPIO issue with long double etypes.
- Update treematch topology component to fix numerous correctness issues.
- Fix memory leak in UCX MCA parameter registration.
- Fix long operation closing file descriptors on non-Linux systems that
  can appear as a hang to users.
- Fix for attribute handling on GCC 11 due to pointer aliasing.
- Fix multithreaded race in UCX PML's datatype handling.

3rd-party/prrte/src/tools/prte_info/prte_info.c
/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2006 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2007 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2007-2020 Cisco Systems, Inc.  All rights reserved
 * Copyright (c) 2010-2016 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2014-2020 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Research Organization for Information Science
 *                         and Technology (RIST). All rights reserved.
 * Copyright (c) 2021-2024 Nanook Consulting  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 */

#include "prte_config.h"

#include <ctype.h>
#include <stdio.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#    include <unistd.h>
#endif
#ifdef HAVE_NETDB_H
#    include <netdb.h>
#endif
#ifdef HAVE_SYS_PARAM_H
#    include <sys/param.h>
#endif
#include <errno.h>
#include <signal.h>

#include "src/class/pmix_object.h"
#include "src/class/pmix_pointer_array.h"
#include "src/mca/base/pmix_base.h"
#include "src/mca/errmgr/errmgr.h"
#include "src/mca/prteinstalldirs/prteinstalldirs.h"
#include "src/mca/schizo/base/base.h"
#include "src/prted/pmix/pmix_server.h"
#include "src/util/pmix_argv.h"
#include "src/util/pmix_basename.h"
#include "src/util/pmix_cmd_line.h"
#include "src/util/error.h"
#include "src/util/pmix_path.h"
#include "src/util/proc_info.h"
#include "src/util/pmix_show_help.h"

#include "constants.h"
#include "src/include/prte_frameworks.h"
#include "src/include/version.h"
#include "src/runtime/prte_locks.h"

#include "src/tools/prte_info/pinfo.h"

/*
 * Public variables
 */

bool prte_info_pretty = true;
pmix_cli_result_t prte_info_cmd_line = PMIX_CLI_RESULT_STATIC_INIT;

const char *prte_info_type_all = "all";
const char *prte_info_type_prte = "prte";
const char *prte_info_type_base = "base";

pmix_pointer_array_t mca_types = {{0}};

int main(int argc, char *argv[])
{
    int ret = 0;
    bool acted = false;
    bool want_all = false;
    int i;
    char *str;
    int option_index = 0;   /* getopt_long stores the option index here. */
    char *ptr;
    char *personality;
    prte_schizo_base_module_t *schizo;

    PRTE_HIDE_UNUSED_PARAMS(argc);

    /* protect against problems if someone passes us thru a pipe
     * and then abnormally terminates the pipe early */
    signal(SIGPIPE, SIG_IGN);

    prte_tool_basename = pmix_basename(argv[0]);
    prte_tool_actual = "prte_info";

    /* Initialize the argv parsing stuff */
    if (PRTE_SUCCESS != (ret = prte_init_util(PRTE_PROC_MASTER))) {
        pmix_show_help("help-prte-info.txt", "lib-call-fail", true, "prte_init_util", __FILE__,
                       __LINE__, NULL);
        exit(ret);
    }

    /* open the SCHIZO framework */
    ret = pmix_mca_base_framework_open(&prte_schizo_base_framework,
                                       PMIX_MCA_BASE_OPEN_DEFAULT);
    if (PRTE_SUCCESS != ret) {
        PRTE_ERROR_LOG(ret);
        return ret;
    }

    if (PRTE_SUCCESS != (ret = prte_schizo_base_select())) {
        PRTE_ERROR_LOG(ret);
        return ret;
    }

    /* look for any personality specification */
    personality = NULL;
    for (i = 0; NULL != argv[i]; i++) {
        if (0 == strcmp(argv[i], "--personality")) {
            personality = argv[i + 1];
            break;
        }
    }

    /* detect if we are running as a proxy and select the active
     * schizo module for this tool */
    schizo = prte_schizo_base_detect_proxy(personality);
    if (NULL == schizo) {
        pmix_show_help("help-schizo-base.txt", "no-proxy", true, prte_tool_basename, personality);
        return 1;
    }
    if (NULL == personality) {
        personality = schizo->name;
    }

    /* Register all global MCA Params */
    if (PRTE_SUCCESS != (ret = prte_register_params())) {
        if (PRTE_ERR_SILENT != ret) {
            pmix_show_help("help-prte-runtime", "prte_init:startup:internal-failure", true,
                           "prte register params",
                           PRTE_ERROR_NAME(ret), ret);
        }
        return 1;
    }

    /* parse the input argv to get values, including everyone's MCA params */
    PMIX_CONSTRUCT(&prte_info_cmd_line, pmix_cli_result_t);
    ret = schizo->parse_cli(argv, &prte_info_cmd_line, PMIX_CLI_SILENT);
    if (PRTE_SUCCESS != ret) {
        PMIX_DESTRUCT(&prte_info_cmd_line);
        if (PRTE_OPERATION_SUCCEEDED == ret) {
            return PRTE_SUCCESS;
        }
        if (PRTE_ERR_SILENT != ret) {
            fprintf(stderr, "%s: command line error (%s)\n", prte_tool_basename, prte_strerror(ret));
        }
        return ret;
    }
    // we do NOT accept arguments other than our own
    if (NULL != prte_info_cmd_line.tail) {
        str = PMIX_ARGV_JOIN_COMPAT(prte_info_cmd_line.tail, ' ');
        if (0 != strcmp(str, argv[0])) {
            ptr = pmix_show_help_string("help-pterm.txt", "no-args", false,
                                        prte_tool_basename, str, prte_tool_basename);
            free(str);
            if (NULL != ptr) {
                printf("%s", ptr);
                free(ptr);
            }
            return -1;
        }
        free(str);
    }

    /* setup the mca_types array */
    PMIX_CONSTRUCT(&mca_types, pmix_pointer_array_t);
    pmix_pointer_array_init(&mca_types, 256, INT_MAX, 128);

    /* add a type for prte itself */
    pmix_pointer_array_add(&mca_types, "mca");
    pmix_pointer_array_add(&mca_types, "prte");

    /* add a type for hwloc */
    pmix_pointer_array_add(&mca_types, "hwloc");

    /* let the pmix server register params */
    pmix_server_register_params();
    /* add those in */
    pmix_pointer_array_add(&mca_types, "pmix");

    /* add the rml and routed types since they are no
     * longer in a framework */
    pmix_pointer_array_add(&mca_types, "rml");
    pmix_pointer_array_add(&mca_types, "routed");

    /* push all the types found by autogen */
    for (i = 0; NULL != prte_frameworks[i]; i++) {
        pmix_pointer_array_add(&mca_types, prte_frameworks[i]->framework_name);
    }

    /* Execute the desired action(s) */
    want_all = pmix_cmd_line_is_taken(&prte_info_cmd_line, "all");
    if (want_all) {
        prte_info_do_version(want_all);
        acted = true;
    } else if (pmix_cmd_line_is_taken(&prte_info_cmd_line, "show-version")) {
        prte_info_do_version(false);
        acted = true;
    }
    if (want_all || pmix_cmd_line_is_taken(&prte_info_cmd_line, "path")) {
        prte_info_do_path(want_all);
        acted = true;
    }
    if (want_all || pmix_cmd_line_is_taken(&prte_info_cmd_line, "arch")) {
        prte_info_do_arch();
        acted = true;
    }
    if (want_all || pmix_cmd_line_is_taken(&prte_info_cmd_line, "hostname")) {
        prte_info_do_hostname();
        acted = true;
    }
    if (want_all || pmix_cmd_line_is_taken(&prte_info_cmd_line, "config")) {
        prte_info_do_config(true);
        acted = true;
    }
    if (want_all || pmix_cmd_line_is_taken(&prte_info_cmd_line, "param")) {
        prte_info_do_params(want_all, pmix_cmd_line_is_taken(&prte_info_cmd_line, "internal"));
        acted = true;
    }

    /* If no command line args are specified, show default set */

    if (!acted) {
        prte_info_show_prte_version(prte_info_ver_full);
        prte_info_show_path(prte_info_path_prefix, prte_install_dirs.prefix);
        prte_info_do_arch();
        prte_info_do_hostname();
        prte_info_do_config(false);
        prte_info_components_open();
        for (i = 0; i < mca_types.size; ++i) {
            if (NULL == (str = (char *) pmix_pointer_array_get_item(&mca_types, i))) {
                continue;
            }
            prte_info_show_component_version(str, prte_info_component_all, prte_info_ver_full,
                                             prte_info_type_all);
        }
    }

    /* All done */
    prte_info_components_close();
    PMIX_DESTRUCT(&mca_types);
    pmix_mca_base_close();

    return 0;
}
